### DISCUSSION TITLE
Flip the column direction for RTL language

### DISCUSSION THREAD

**Discussion by nakibrayane** [YaLTeR/niri#658] [Ideas]
This may be hard to implement, and I would understand if your wouldn't implement this.

User interfaces in RTL languages should be flipped. it makes more seance if the column would scroll from right to left, meaning that new windows would spawn in the left of the current window. and all keybindings involving moving in the column, or moving windows inside the column should be flipped.

it's weird for me that windows spawn in right of the current one.

see nautilus first in RTL, and second in LTR.

![20240907_100841](https://github.com/user-attachments/assets/d8f39535-df3d-4414-9342-2f10381a6f90)

![20240907_101014](https://github.com/user-attachments/assets/35f15d3d-bb49-4662-92c7-7d05af70dc15)

**Comment by YaLTeR** [comment]
Yeah, I know about this behavior. I'm however not entirely sure yet how to implement this. Niri doesn't do any localization yet and I'm assuming RTL is a property of the current language? 

Also can't say how hard it would be code wise. 

What do you mean by bind flipping? The binds are directional (i.e. focus-column-left and it's bound to the left arrow), why should they be flipped?

**Reply by nakibrayane** [reply]
> Yeah, I know about this behavior. I'm however not entirely sure yet how to implement this. Niri doesn't do any localization yet and I'm assuming RTL is a property of the current language?

Yes it is a property of the language. if the language is written from right to left the ui should also be in RTL.

And it would be nice if i could translate the text on the screenshot and quit niri overlays.

> Also can't say how hard it would be code wise.

I can't comment on this, my coding skills in rust are close to zero :(

> What do you mean by bind flipping? The binds are directional (i.e. focus-column-left and it's bound to the left arrow), why should they be flipped?

Yes, you are right, maybe the bindings shouldn't be flipped.

**Reply by YaLTeR** [reply]
This is how GTK picks the text direction: https://gitlab.gnome.org/GNOME/gtk/-/blob/72718b719387d15a29bf4524bc721eec490b357e/gtk/gtkmain.c#L774

It enumerates the scripts for the current locale with Pango, then asks Harfbuzz which text direction they use until it finds an LTR or and RTL script.

**Reply by nakibrayane** [reply]
> This is how GTK picks the text direction: https://gitlab.gnome.org/GNOME/gtk/-/blob/72718b719387d15a29bf4524bc721eec490b357e/gtk/gtkmain.c#L774
> 
> It enumerates the scripts for the current locale with Pango, then asks Harfbuzz which text direction they use until it finds an LTR or and RTL script.

I think this is used for a line of text or paragraph that has both RTL and LTR script (which is common), And it has to order each script correctly.

In the case of niri this is not usefull. you can just read $LANG environment variable and match the languages if their script is written from right to left then reverse the column.

Or a better solution would be to add an option, for example: `reverse-column-direction`.

**Reply by nakibrayane** [reply]
Or an option to set the column direction in eather: LTR, RTL, TTB BTT.

I read a lot of requests in this repo to have a vertical column stack, this would solve those issues too.

**Reply by YaLTeR** [reply]
> you can just read $LANG environment variable and match the languages if their script is written from right to left

Yes, that is what the GTK code that I linked does.

**Reply by zefr0x** [reply]
> This is how GTK picks the text direction: https://gitlab.gnome.org/GNOME/gtk/-/blob/72718b719387d15a29bf4524bc721eec490b357e/gtk/gtkmain.c#L774
> 
> It enumerates the scripts for the current locale with Pango, then asks Harfbuzz which text direction they use until it finds an LTR or and RTL script.

I don't think this is necessary.

Niri is not a full desktop environment like Gnome or KDE, the user is expected to set a config option for this behaviour.

**Reply by nakibrayane** [reply]
I agree, this should be an option.

**Comment by snoopdouglas** [comment]
Also mentioned on #1159, but I'll put it here too: I'm a native English speaker, but I've got bad vision in my right eye, which means I tend to want the most 'current' or 'recent' task on the left of the screen. So this'd be a good fit for my use case too.

**Comment by haasn** [comment]
Tangentially related, but in a two monitor setup I would like to have the tiling direction flipped for the left monitor only, so that the windows on either side of the vertical split are growing "outwards" from the center.

e.g.
```
[      3 2 1 ][ 1 2 3      ]
```

**Reply by funme** [reply]
Has this feature been implemented?

**Reply by AlpSha** [reply]
I'm looking for this too

**Comment by nakibrayane** [comment]
An option to change the column scroll direction would be vey useful in many situations.

**Comment by markstos** [comment]
The conversation in https://github.com/YaLTeR/niri/discussions/1642 is requesting the same thing, for different reasons: Some people with 3 monitors want the windows on the left monitor to be RTL aligned so that windows appear more toward the center.

**Comment by markstos** [comment]
@funme I have a draft PR which adds `anchor "right"`  layout option https://github.com/YaLTeR/niri/pull/2998

**Comment by yngvildr-the-voracious** [comment]
I came and looked for RTL scrolling when I opened a tab looking for family stuff

I see a PR has been opened on december but I'll give my two cents on UX : 

I only searched "right to left" in github issues and did not encounter it, which might be extremely lazy. Since I don't think it is in the config.kdl example file (or if it is I missed it, sorry) I'm posting here ???


**Feature idea** : Right to Left scrolling mode. 

**Why** : Because RTL writing systems condition the brains to look for important info to the right rather than the left for LTR writing systems users. RTL writing users might find the feature a very interesting QoL option. I'm talking out of my French ass, tho. 
*Edit while reading the present discussion thread* : other demographics find it useful for accessibility or workflow reasons, widening the pool or arguments for the feature. 

**How (UX)** : config.kdl option, ideally placed around the keyboard layout options for ease of use, ease of search and ease of locale setup + software coded to edit config files would be able to find it 

**How (Rust)** : I can only speak basic computer, nothing beyond "pacman -Syu", "gitignore secret-diary.rtf" and "sudo apt install --fix-broken", sorry. I don't even speak Arabic beyond a few dz-ar swears, I just had an Algerian mom and way too much interest in linguistics and brain developement. (Though i am teaching myself rust using cosmic-comp's code base and an LLM chat to guide me on how it works so I can shift it to a scrolling rather than tree, so... Ask me again in a century, I guess?)
*Edit while reading the present discussion thread* : @markstos doing the FOSS Gods work in #2998 with his PoC. May you ascend to free open software Valhalla. 

**How it could evolve into if niri takes over the world and becomes your mom's WM :** If you dev it with enough of a macro view on the code, you might scroll up to top and top to down but I'm not sure what kind of demographic that would serve right now and open source means lots of work and few monies in the pocket. Could be someone's vanity project, tho? I don't have said macro view on the code (or even just coding beyond a few tweaks and scripts), meaning that it could take an entire unloading of the yarnball type of full scale rewrite, but it's FOSS, forks happen. 

شكرا لك
for listening :)
